### PR TITLE
Bug/rebalancing

### PR DESCRIPTION
**Pull request by Maxtaager**

### CODE REVIEW

File: motorway/contrib/amazon_kinesis/ramps.py [hunk 1]
```diff
@@ -97,13 +97,25 @@ def claim_shard(self, shard_id):
...
+        control_record = None
+        shards_by_shard_id = {}
+        for shard in shards:
+            if shard['shard_id'] == shard_id:
+                control_record = shard
+            # make a copy
+            shards_by_shard_id[shard['shard_id']] = dict(shard)
+
+        if control_record is None:
+            return NoItemsReturned()
```
Comment (KristianOellegaard) [inline motorway/contrib/amazon_kinesis/ramps.py]: IS this right?

Comment (KristianOellegaard) [reply motorway/contrib/amazon_kinesis/ramps.py]: Should probably have a test

File: motorway/contrib/amazon_kinesis/ramps.py [hunk 2]
```diff
@@ -120,7 +132,7 @@ def can_claim_shard(self, shard_id):
         for shard in shards:  # Update active worker cache
             if shard['worker_id'] not in active_workers:
                 heartbeats[shard['worker_id']] = shard['heartbeat']
-            if heartbeats[shard['worker_id']] == shard['heartbeat']:
+            if shard['heartbeat'] == shards_by_shard_id[shard['shard_id']]['heartbeat']:
```
Comment (KristianOellegaard) [inline motorway/contrib/amazon_kinesis/ramps.py]: What does this change do?

File: motorway/contrib/amazon_kinesis/ramps.py [hunk 3]
```diff
@@ -97,13 +97,25 @@ def claim_shard(self, shard_id):
...
+
         heartbeat = control_record['heartbeat']
+        worker_id = control_record['worker_id']
         time.sleep(self.heartbeat_timeout)
-        if heartbeat == self.control_table.get_item(Key={'shard_id': shard_id})['Item']['heartbeat']:  # TODO: check worker_id as well
+        updated_control_record = self.control_table.get_item(Key={'shard_id': shard_id})['Item']
+
+        if heartbeat == updated_control_record['heartbeat'] and updated_control_record['worker_id'] == worker_id:
+            # if both the heartbeat and the worker_id is the same
             shard_election_logger.debug("Shard %s - heartbeat remained unchanged for defined time, taking over" % shard_id)
```
Comment (KristianOellegaard) [inline motorway/contrib/amazon_kinesis/ramps.py]: Should say "heartbeat and worker id remained..."

Comment (KristianOellegaard) [reply motorway/contrib/amazon_kinesis/ramps.py]: Maybe more logging for other cases

File: motorway/contrib/amazon_kinesis/ramps.py (hunk 1 above)
Comment (Maxtaager) [reply motorway/contrib/amazon_kinesis/ramps.py]: fixed and added test

File: motorway/contrib/amazon_kinesis/ramps.py (hunk 2 above)
Comment (Maxtaager) [reply motorway/contrib/amazon_kinesis/ramps.py]: made it more readable by only using 1 dictionary with heartbeats.

File: motorway/contrib/amazon_kinesis/ramps.py (hunk 3 above)
Comment (Maxtaager) [reply motorway/contrib/amazon_kinesis/ramps.py]: fixed and added logging for when the worker_id changes.

**Comment by KristianOellegaard** [comment]
LGTM

**Pull request merged**